runtime/oti/j9accessbarrier.h

```diff
@@ -113,12 +113,32 @@ typedef struct J9IndexableObject* mm_j9array_t;
 		? (&(((elemType*)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, ((U_32*)(((UDATA)(array)) + (javaVM)->discontiguousIndexableHeaderSize))[((U_32)index)/((javaVM)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%((javaVM)->arrayletLeafSize/sizeof(elemType))])) \
 		: (&(((elemType*)(((UDATA*)(((UDATA)(array)) + (javaVM)->discontiguousIndexableHeaderSize))[((U_32)index)/((javaVM)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%((javaVM)->arrayletLeafSize/sizeof(elemType))])))
 
+#if defined(J9VM_ENV_DATA64)
+#define J9JAVAARRAYCONTIGUOUS_EA(vmThread, array, index, elemType) \
+		((vmThread->isIndexableDataAddrPresent) \
+			?	(J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread) \
+				? (&((elemType*)((((J9IndexableObjectWithDataAddressContiguousCompressed *)(array))->dataAddr)))[(index)]) \
+				: (&((elemType*)((((J9IndexableObjectWithDataAddressContiguousFull *)(array))->dataAddr)))[(index)])) \
+			: (&((elemType*)((((UDATA)(array)) + (vmThread)->contiguousIndexableHeaderSize)))[(index)]))
```

Comment:
it would be slightly more optimal to rely on compile time sizes of headers for the case when dataAddr is not present  (basically rely on same macros pre-dual header changes)

indexableHeaderSize fields are still good to use for other non-perf critical code, but here we prefer performance over not-so-readable code

Reply:
to slightly improve readability of this big macro, perhaps makes sense to introduce a couple of more specialized sub-macros: contiguous EA with/without dataAddr? if so, the one for without dataAddr, could be used also in 32bit case

Reply:
Good idea. This will definitely improve readability.